src/modules/roc_audio/resampler_map.cpp

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright (c) 2019 Roc authors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include "roc_audio/resampler_map.h"
+#include "roc_audio/resampler_builtin.h"
+#include "roc_audio/resampler_config.h"
+
+namespace roc {
+namespace audio {
+
+IResampler* ResamplerMap::new_resampler(ResamplerBackend resampler_backend,
+                                    core::IAllocator& allocator,
+                                    const ResamplerConfig& config,
+                                    packet::channel_mask_t channels,
+                                    size_t frame_size){
+    
+    switch (resampler_backend)
+    {
+    case BUILTIN_RESAMPLER:
+        current_resampler.reset(new (allocator) BuiltinResampler(allocator, config, channels, frame_size), allocator);
```

Comment:
I also suggest to remove `valid()` from IResampler interface and check it here, before returning the pointer, like:

```
UniquePtr<BuiltinResampler> r(new (allocator) BuiltinResampler());
if (!r || !r->valid()) {
    return NULL;
}
return r.release();
```

This approach will give a guarantee that ResamplerMap may return only successfully constructed objects, which would be nice. We already use such approach in other factories.

Reply:
> It's a bit surprising that `new_resampler()` changes the state of ResamplerMap and even destroys the previously created resampler. It is also dangerous given that we return a raw pointer to the user. If the user will store the pointer and then call `new_resampler()` again, the pointer will become invalid.
> 
> We also have `get_current_resampler()`, but it seems it's not used anywhere.
> 
> I suggest to make ResamplerMap stateless and remove `current_resampler` and `get_current_resampler()`. It's common for factory pattern.
> 
> `new_resampler()` can just return a raw pointer to the resampler and the user will be responsible to store it into UniquePtr. It will be allowed to share ResamplerMap instance and call `new_resampler()` to create several resamplers simultaneously.

You are right, I assumed there that every user would instantiate ResamplerMap to then use a concrete resampler and give them the responsibility to use the IResampler pointer well, but it might be confusing to use in that way. I made the change that you suggested.

> I also suggest to remove `valid()` from IResampler interface and check it here, before returning the pointer, like:
> 
> ```
> UniquePtr<BuiltinResampler> r(new (allocator) BuiltinResampler());
> if (!r || !r->valid()) {
>     return NULL;
> }
> return r.release();
> ```
> 
> This approach will give a guarantee that ResamplerMap may return only successfully constructed objects, which would be nice. We already use such approach in other factories.

I think it would be more convenient to keep the valid() method inside IResampler to avoid doing the check with every backend and instead doing it once after the switch statement.

Reply:
> I think it would be more convenient to keep the valid() method inside IResampler to avoid doing the check with every backend and instead doing it once after the switch statement.

Makes sense. Let's keep it.